mods/stairs/init.lua

```diff
@@ -129,86 +134,60 @@ function stairs.register_slab(subname, recipeitem, groups, images, description,
-				local success
-				fakestack, success = minetest.item_place(fakestack, placer,
-					pointed_thing)
-				-- If the item was taken from the fake stack, decrement original
-				if success then
-					itemstack:set_count(fakestack:get_count())
-				-- Else put old node back
-				else
-					minetest.set_node(slabpos, slabnode)
-				end
-				return itemstack
+			local rot = slab_trans_dir_place[dir]
+			if rot == 0 or rot == 20 then
+				rot = rot + minetest.dir_to_facedir(placer:get_look_dir())
 			end
-			
-			-- Upside down slabs
-			if p0.y - 1 == p1.y then
-				-- Turn into full block if pointing at a existing slab
-				if n0_is_upside_down  then
-					-- Remove the slab at the position of the slab
-					minetest.remove_node(p0)
-					-- Make a fake stack of a single item and try to place it
-					local fakestack = ItemStack(recipeitem)
-					fakestack:set_count(itemstack:get_count())
-
-					pointed_thing.above = p0
-					local success
-					fakestack, success = minetest.item_place(fakestack, placer,
-						pointed_thing)
-					-- If the item was taken from the fake stack, decrement original
-					if success then
-						itemstack:set_count(fakestack:get_count())
-					-- Else put old node back
-					else
-						minetest.set_node(p0, n0)
+
+			-- place_node will in turn attempt to call on_rightclick()
+			return minetest.item_place(itemstack, placer, pointed_thing, rot)
+		end,
+		on_rightclick = function(pos, node, clicker, itemstack, pointed_thing)
+			local under = minetest.get_node(pointed_thing.under)
+			local dir = minetest.dir_to_facedir(vector.subtract(
+				pointed_thing.above, pointed_thing.under), true)
+
+			local wield_item = clicker:get_wielded_item():get_name()
+			local player_name = clicker:get_player_name()
+			local p2 = node.param2
+
+			if wield_item == node.name then
+				-- combine two slabs
+				if slab_trans_dir[math.floor(p2 / 4)] == dir then
+					-- combinable slab?
+					if not recipeitem then
+						return itemstack
+					end
+					if minetest.is_protected(pos, player_name) and not
+							minetest.check_player_privs(placer, "protection_bypass") then
+						minetest.record_protection_violation(pos, player_name)
+						return
+					end
+					minetest.set_node(pos, {name = recipeitem, param2 = p2})
+					if not minetest.setting_getbool("creative_mode") then
+						itemstack:take_item()
 					end
 					return itemstack
 				end
 
-				-- Place upside down slab
-				param2 = 20
+				-- Placing a slab on an upside down slab should make it right-side up.
+				if p2 >= 20 and dir == 8 then
+					p2 = p2 - 20
+				-- same for the opposite case: slab below normal slab
+				elseif p2 <= 3 and dir == 4 then
+					p2 = p2 + 20
+				end
 			end
 
-			-- If pointing at the side of a upside down slab
-			if n0_is_upside_down and p0.y + 1 ~= p1.y then
-				param2 = 20
+			-- else attempt to place node with proper param2
+			minetest.item_place_node(ItemStack(wield_item), clicker, pointed_thing, p2)
```

Comment:
Why not just return `minetest.item_place_node`'s return value? The following code seems redundant.